### DISCUSSION TITLE
Transient instability in high flowrate model

### DISCUSSION THREAD

**Discussion by jvwilliams23** [idaholab/moose#25947] [Q&A Modules: Thermal Hydraulics] [answered]
Hi,

I have spent a couple of months now trying to deal with transient instability in a model with high flowrate (need to go up to 1000 L/min, giving an inlet velocity as 140 m/s). I have simplified the geometry significantly and created an MWE which reproduces the issue. The figure below shows the pressure drop increasing exponentially from about `t=0.05 s` to eventually crash at `t~1.61 s`.
![pressure-crash](https://github.com/idaholab/moose/assets/48445365/bfe85cc1-4758-4f0f-aa4b-8ad7c0d3207f)

Following the guidelines in ["Troubleshooting failed solves"](https://mooseframework.inl.gov/application_usage/failed_solves.html) page, I have changed the initial condition from `velocity=0` to one based on inlet velocity and flow divisions, I also fixed the diameter as constant. I also scaled the variables to improve the solution. The convergence seems fine throughout the whole simulation (see log below):

```
Time Step 16100, time = 1.61, dt = 0.0001
    |residual|_2 of individual variables:
         junction-728:rhoEV: 0.954486
         junction-728:rhoV:  0.00838734
         junction-728:rhouV: 42.7522
         junction-728:rhovV: 19.5515
         junction-728:rhowV: 4.37464
         junction-733:rhoEV: 5.49613
         junction-733:rhoV:  0.00634548
         junction-733:rhouV: 24.375
         junction-733:rhovV: 110.444
         junction-733:rhowV: 288.872
         rhoA:               0.0411604
         rhoEA:              0.0550817
         rhouA:              0.060198
 0 Nonlinear |R| = 3.138465e+02
      0 Linear |R| = 3.138465e+02
      1 Linear |R| = 2.163226e+02
      2 Linear |R| = 1.535664e+00
      3 Linear |R| = 2.392176e-01
      4 Linear |R| = 5.622355e-03
      5 Linear |R| = 1.110909e-05
    |residual|_2 of individual variables:
         junction-728:rhoEV: 0.0501407
         junction-728:rhoV:  3.20005e-05
         junction-728:rhouV: 1.77731
         junction-728:rhovV: 1.16674
         junction-728:rhowV: 0.00863554
         junction-733:rhoEV: 0.54433
         junction-733:rhoV:  0.00175378
         junction-733:rhouV: 1.42963
         junction-733:rhovV: 9.53675
         junction-733:rhowV: 0.163855
         rhoA:               0.000152264
         rhoEA:              0.000190096
         rhouA:              0.000191392
 1 Nonlinear |R| = 9.891371e+00
      0 Linear |R| = 9.891371e+00
      1 Linear |R| = 6.471072e-02
      2 Linear |R| = 8.884682e-03
      3 Linear |R| = 3.197219e-04
      4 Linear |R| = 2.780930e-07
    |residual|_2 of individual variables:
         junction-728:rhoEV: 3.04345e-06
         junction-728:rhoV:  7.24824e-09
         junction-728:rhouV: 2.56967e-06
         junction-728:rhovV: 4.03091e-06
         junction-728:rhowV: 9.483e-09
         junction-733:rhoEV: 0.000150371
         junction-733:rhoV:  7.11678e-07
         junction-733:rhouV: 0.000148026
         junction-733:rhovV: 0.000999644
         junction-733:rhowV: 1.02263e-05
         rhoA:               6.19741e-08
         rhoEA:              1.52932e-07
         rhouA:              8.7553e-08
 2 Nonlinear |R| = 1.021738e-03
      0 Linear |R| = 1.021738e-03
      1 Linear |R| = 1.061215e-05
      2 Linear |R| = 4.211359e-07
      3 Linear |R| = 1.911311e-08
      4 Linear |R| = 3.632826e-12
    |residual|_2 of individual variables:
         junction-728:rhoEV: 2.11386e-13
         junction-728:rhoV:  1.88877e-14
         junction-728:rhouV: 4.01883e-11
         junction-728:rhovV: 4.57945e-12
         junction-728:rhowV: 1.66978e-13
         junction-733:rhoEV: 1.39089e-12
         junction-733:rhoV:  2.88641e-14
         junction-733:rhouV: 1.24345e-13
         junction-733:rhovV: 1.33582e-12
         junction-733:rhowV: 7.95808e-13
         rhoA:               1.56777e-12
         rhoEA:              2.15145e-12
         rhouA:              2.16606e-12
 3 Nonlinear |R| = 4.064838e-11
 Solve Converged!

Outlier Variable Residual Norms:
  junction-728:rhouV: 4.018830e-11
```

But not long after, the time-step becomes very small and the simulation crashes. I thought maybe this was due to the fluid properties which I fixed as constant becoming less accurate as pressure and temperature change non-homogeneously, but when I changed to Water97FluidProperties the simulation diverges really quick (within a few steps due to issue with line search), so I guess that is a different problem and I should focus on this one first.

I have attached the input file. Hopefully someone could manage to provide some assistance. Happy to provide more information, log files or anything else that could be useful.

Best wishes,
Josh
[coolant-simulation-for-github.i.txt](https://github.com/idaholab/moose/files/13236320/coolant-simulation-for-github.i.txt)

**Comment by jvwilliams23** [comment]
Seems that when I increase the volume at a VolumeJunction from 9e-7 to 9e-4 (just a random test) that everything is stable. Not sure if this is the underlying or is maybe just masking it (something to do with variable scaling in junctions).

How should the volume be calculated? In my input files I have been treating the junction as a sphere with the radius equal to that of the upstream branch, and calculating volume of the sphere (which gave 9e-7).

**Reply by jvwilliams23** [reply]
Increasing from 9e-7 to 9e-6 seems to delay the issue, but still arises after a while.
![pressure-crash-vol9e-6](https://github.com/idaholab/moose/assets/48445365/269e7479-812b-49a1-806c-ab53be706425)

**Reply by GiudGiud** [reply]
@joshuahansel

**Comment by joshuahansel** [comment]
I tried your input file without the loop 728-784-791-733, and it finishes fine, so it's the loop causing issues. In particular, I think the formulation of `VolumeJunction1Phase` needs improvement. We've had issues with this junction in the past but have not had the funding to tackle it, despite it being a pretty critical component. The only other junction that can handle multiple connections is `JunctionParallelChannels1Phase`, but this of course requires that channels be parallel, which is not your situation. I'm going to try and look at the volume junction formulation soon, but I can't guarantee a solution soon.

**Reply by joshuahansel** [reply]
I found something that I can try, so I'm going to try it now. You should hopefully hear from me by the end of the day.

**Reply by joshuahansel** [reply]
Well I implemented the change (I think correctly...), but I get a lot of nonlinear convergence difficulty on your test problem, even after exploring it for a few hours. At this point, I think I would need to properly fund the task because apparently it is not a quick fix. I made an issue here: https://github.com/idaholab/moose/issues/25958. I will see if I can get this work funded.

**Reply by jvwilliams23** [reply]
Yes, the loop is the problem I found too. Thanks for your time.

**Reply by joshuahansel** [reply]
To answer your question about junction volume, yes, that's how we typically do it. You could also choose a sphere with the same surface area as the sum of the connected cross-sectional areas - maybe that would be better. A tip: you can display the junction variables by setting `Outputs/disable_scalars_in_console=false`. One thing that was strange that I saw in my experimenting was that the variable `rhoV` was going up significantly (in a relative sense) in some cases.

In case it's useful, here's an input file with the same general setup:
[bypass_loop.txt](https://github.com/idaholab/moose/files/13251582/bypass_loop.txt)
It appears to give a good solution. You can modify it gradually toward your input file to see what particular thing causes things to break down. That would be helpful to us (maybe to you too).

**Reply by joshuahansel** [reply]
See the issue referenced above for an update. I will try the suggested change and see if this improves things.

**Reply by jvwilliams23** [reply]
Thanks for the answer and the case setup file.
Interesting point raised in the issue regarding the tangential momentum. When I was setting up the MWE, I found that if I remove the junctions "junction-232" and "junction-14" (pretty much the bends in the pipe attached to inlet and outlet, which would create a straight path from the inlet to the volumejunction, and volumejunction to outlet) then the simulation was stable. I could not figure out why, so I guess this is maybe related.

**Reply by joshuahansel** [reply]
Good news: I've reworked the volume junction and I tried your input file again, and it ran well. Got to steady state at 1.25 seconds - took 11 time steps. Solution looks plausible.

I have one more thing to try out on it, and then I'll put up a merge request.

**Reply by jvwilliams23** [reply]
Great news, thanks! I look forward to testing it.

**Reply by joshuahansel** [reply, marked as answer]
I'm getting waylaid by other work, and this will take some work like documentation and testing before it can be reviewed and merged, but I put up the merge request as a draft if you urgently need the branch for your work: https://github.com/idaholab/moose/pull/26046.

**Reply by jvwilliams23** [reply]
I have tested it on the MWE and the more complex case, both seem to be working well. Thanks a lot for your help! I will mark this as answered.
